### ISSUE TITLE
A Map of Everywhere My Family Has Ever Been

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#4540]
A Map of Everywhere My Family Has Ever Been<br>
<br>
http://bit.ly/2IyRMMJ<br>
<br>
Logo
Created using Figma<br>
<br>
<div><article>
<p><img alt="travelmap" src="http://jamie-wong.com/images/14-01-03/travelmap.png"></p>
<p><em>This is the background and tech teardown of making my <a href="http://jamie-wong.com/experiments/travelmap/">Travel Map</a>, which
you should check out before reading this post. You can find the full source on
github at <a href="https://github.com/jlfwong/travelmap">jlfwong/travelmap</a>.</em></p>
<nav>
<ul>
<li><a href="#background">Background</a></li>
<li><a href="#data-format">Data Format</a></li>
<li><a href="#geocoding">Geocoding</a></li>
<li>
<a href="#promises">Promises</a>
<ul>
<li><a href="#transforming-asynchronous-results">Transforming Asynchronous Results</a></li>
<li><a href="#serial-requests">Serial Requests</a></li>
<li><a href="#parallel-requests">Parallel Requests</a></li>
</ul>
</li>
<li>
<a href="#rendering">Rendering</a>
<ul>
<li><a href="#projections">Projections</a></li>
<li><a href="#places-and-paths">Places and Paths</a></li>
<li><a href="#animating">Animating</a></li>
</ul>
</li>
<li><a href="#brunch">Brunch</a></li>
<li><a href="#caching-in-localstorage">Caching in LocalStorage</a></li>
</ul>
</nav>
<p>After I ended my 11 week tour of Western Europe, I felt it was my nerdy
obligation to create some form of data visualization for the trip. I’d also been
looking for an excuse to learn <a href="http://d3js.org/">D3.js</a> for some time. As a third stroke of
luck, my mom just broke her foot (bear with me, I swear I’m not a terrible son).
She was housebound and I know she absolutely <em>hates</em> being without things to
keep her busy, so I gave her a gargantuan task. I asked her to help me collect a
list of every city everyone in my immediate family had ever slept in, in order,
including returning home (so Ottawa → Toronto → Ottawa).</p>
<p>While she was poring over old diaries and calendars, I got started reading. For
getting a good overview on technical subjects now, I prefer starting with books
over tutorials or blog posts because they tend to do a better job of giving me
vocabulary to work with. After a brief look around, I settled upon <a href="http://www.tkqlhce.com/click-7365820-11260198?url=http%3A%2F%2Fshop.oreilly.com%2Fproduct%2F0636920026938.do%3Fcmp%3Daf-strata-books-videos-product_cj_9781449339722_%2525zp&amp;cjsku=9781449339722">Interactive
Data Visualization for the Web</a>, which I thoroughly enjoyed.</p>
<p>I knew the data entry for this was going to be time consuming, and I’d be
getting data in different formats from each of my family members, so I decided
to opt for something simple. The data file is a JavaScript file which exports a
mapping from the person’s name to the list of places they’ve been in order. So a
(very) reduced version would look like this:</p>
<div><pre><code data-lang="javascript"><span></span><span>module</span><span>.</span><span>exports</span> <span>=</span> <span>{</span>
  <span>"Jamie"</span><span>:</span> <span>[</span><span>"Ottawa"</span><span>,</span> <span>"Waterloo"</span><span>,</span> <span>"Toronto"</span><span>,</span> <span>"San Francisco"</span><span>,</span> <span>"Ottawa"</span><span>]</span>
  <span>"Tammy"</span><span>:</span> <span>[</span><span>"Ottawa"</span><span>,</span> <span>"Toronto"</span><span>,</span> <span>"Ottawa"</span><span>],</span>
  <span>"Becky"</span><span>:</span> <span>[</span><span>"Ottawa"</span><span>,</span> <span>"Winnipeg"</span><span>,</span> <span>"Ottawa"</span><span>,</span> <span>"Saskatoon"</span><span>],</span>
  <span>"Emma"</span><span>:</span> <span>[</span><span>"Ottawa"</span><span>,</span> <span>"Kingston"</span><span>,</span> <span>"Ottawa"</span><span>,</span> <span>"Vancouver"</span><span>],</span>
  <span>"Susan"</span><span>:</span> <span>[</span><span>"Sheffield, England"</span><span>,</span> <span>"Calgary"</span><span>,</span> <span>"Ottawa"</span><span>],</span>
  <span>"Ging"</span><span>:</span> <span>[</span><span>"Hong Kong"</span><span>,</span> <span>"Calgary"</span><span>,</span> <span>"Ottawa"</span><span>]</span>
<span>}</span>
</code></pre></div>
<p>The places listed are just strings containing human readable place names
(usually cities).</p>
<p>After the data started pouring in, I realized that updating all of our lists for
family trips and having massive data duplication was going to be a pain,
especially for road trips with 10+ stops. I needed some way of saying “all of
these people went on this trip”. I chose the first thing to come to mind: make
each of these family trips an array, then just flatten each person’s list before
it got processed. So to add a shared trip which I went on with my parents and
one of my sisters, it would look like this:</p>
<div><pre><code data-lang="javascript"><span></span><span>var</span> <span>ROADTRIP</span> <span>=</span> <span>[</span>
  <span>"Kalamazoo, Michigan"</span><span>,</span>
  <span>"Chicago, Illinois"</span><span>,</span>
  <span>"Mt Rushmore, South Dakota"</span><span>,</span>
  <span>"Calgary"</span><span>,</span>
  <span>"Windermere, Canada"</span><span>,</span>
  <span>"Waterton, Alberta"</span><span>,</span>
  <span>"Ottawa"</span>
<span>];</span>

<span>module</span><span>.</span><span>exports</span> <span>=</span> <span>{</span>
  <span>"Jamie"</span><span>:</span> <span>[</span><span>"Ottawa"</span><span>,</span> <span>ROADTRIP</span><span>,</span> <span>"Waterloo"</span><span>,</span> <span>"Toronto"</span><span>,</span> <span>"San Francisco"</span><span>,</span> <span>"Ottawa"</span><span>]</span>
  <span>"Tammy"</span><span>:</span> <span>[</span><span>"Ottawa"</span><span>,</span> <span>ROADTRIP</span><span>,</span> <span>"Toronto"</span><span>,</span> <span>"Ottawa"</span><span>],</span>
  <span>"Becky"</span><span>:</span> <span>[</span><span>"Ottawa"</span><span>,</span> <span>"Winnipeg"</span><span>,</span> <span>"Ottawa"</span><span>,</span> <span>"Saskatoon"</span><span>],</span>
  <span>"Emma"</span><span>:</span> <span>[</span><span>"Ottawa"</span><span>,</span> <span>"Kingston"</span><span>,</span> <span>"Ottawa"</span><span>,</span> <span>"Vancouver"</span><span>],</span>
  <span>"Susan"</span><span>:</span> <span>[</span><span>"Sheffield, England"</span><span>,</span> <span>"Calgary"</span><span>,</span> <span>"Ottawa"</span><span>,</span> <span>ROADTRIP</span><span>],</span>
  <span>"Ging"</span><span>:</span> <span>[</span><span>"Hong Kong"</span><span>,</span> <span>"Calgary"</span><span>,</span> <span>"Ottawa"</span><span>,</span> <span>ROADTRIP</span><span>]</span>
<span>};</span>
</code></pre></div>
<p>You can see what the full dataset looks like in <a href="https://github.com/jlfwong/travelmap/blob/master/app/data.js"><code>app/data.js</code></a>.</p>
<p>The data processing phase consists of a pretty normal pattern for data
transformation: a series of <code>map</code>’s (the functional kind, not the geographic
kind – transforming every data point in isolation), followed by a series of
<code>reduce</code>’s (using the result of all the <code>map</code> calls to produce interesting data
and statistics at the end).</p>
<p>In order to plot all these locations on a map, I need to know the corresponding
longitude and latitude for each place. The process of converting human readable
addresses into longitude and latitude is called “geocoding”. Google Maps and
<a href="http://www.openstreetmap.org/about">OpenStreetMap</a> both provide a geocoding service. I opted for the
OpenStreetMap version because it’s very, very simple to use, and I found it gave
me good enough results. An example geocoding query looks like this:</p>
<blockquote>
<p><a href="http://nominatim.openstreetmap.org/search/?q=Paris&amp;format=json">http://nominatim.openstreetmap.org/search/?q=Paris&amp;format=json</a></p>
</blockquote>
<p>I was also interested in some additional information about each place, most
importantly the containing country. Once I have the longitude and latitude of
each place, I can request more information about what exactly exists at those
coordinates by performing “reverse geocoding”. OpenStreetMap conveniently also
provides this as a service. An example reverse geocoding query looks like this:</p>
<blockquote>
<p><a href="http://nominatim.openstreetmap.org/reverse?lat=48.8565056&amp;lon=2.3521334&amp;zoom=8&amp;format=json">http://nominatim.openstreetmap.org/reverse?lat=48.8565056&amp;lon=2.3521334&amp;zoom=8&amp;format=json</a></p>
</blockquote>
<p>You can see the full data processing pipeline in <a href="https://github.com/jlfwong/travelmap/blob/master/app/aggregate.js"><code>app/aggregate.js</code></a></p>
<p>The data retrieval and transformation pipeline here gets pretty complicated, so
I was grateful to make use of Promises. Domenic Denicola does a good job
explaining the benefits of promises in his article <a href="http://domenic.me/2012/10/14/youre-missing-the-point-of-promises/">You’re Missing the Point of
Promises</a>. He also points out the sad mutability of jQuery’s
<a href="http://api.jquery.com/category/deferred-object/"><code>Deferred</code></a> objects, but I thankfully didn’t run into any of those
problems.</p>
<p>I’ll leave you to read that Domenic’s post to argue for the virtues of promises
and focus this section on going over the 3 different ways I used promises to
build the map.</p>
<h2>Transforming Asynchronous Results</h2>
<p>The first is to perform data transformations on the result of an asynchronous
request, as is used in the <code>geocode</code> function. If a <code>then</code> callback returns a
value, then the resulting promise will be immediately resolved with the result.</p>
<div><pre><code data-lang="javascript"><span></span><span>var</span> <span>geocode</span> <span>=</span> <span>function</span> <span>(</span><span>q</span><span>)</span> <span>{</span>
  <span>return</span> <span>$</span><span>.</span><span>ajax</span><span>(</span><span>"http://nominatim.openstreetmap.org/search/"</span><span>,</span> <span>{</span>
    <span>data</span><span>:</span> <span>{</span>
      <span>q</span><span>:</span> <span>q</span><span>,</span>
      <span>format</span><span>:</span> <span>"json"</span>
    <span>}</span>
  <span>}).</span><span>then</span><span>(</span><span>function</span><span>(</span><span>data</span><span>)</span> <span>{</span>
    <span>if</span> <span>(</span><span>!</span><span>data</span> <span>||</span> <span>!</span><span>data</span><span>[</span><span>0</span><span>])</span> <span>{</span>
      <span>throw</span> <span>new</span> <span>Error</span><span>(</span><span>"Geocoding '"</span> <span>+</span> <span>q</span> <span>+</span> <span>"' failed."</span><span>);</span>
    <span>}</span>
    <span>return</span> <span>{</span>
      <span>lat</span><span>:</span> <span>parseFloat</span><span>(</span><span>data</span><span>[</span><span>0</span><span>].</span><span>lat</span><span>,</span> <span>10</span><span>),</span>
      <span>lon</span><span>:</span> <span>parseFloat</span><span>(</span><span>data</span><span>[</span><span>0</span><span>].</span><span>lon</span><span>,</span> <span>10</span><span>)</span>
    <span>};</span>
  <span>});</span>
<span>};</span>
</code></pre></div>
<p>This allows a usage pattern like this:</p>
<div><pre><code data-lang="javascript"><span></span><span>geocode</span><span>(</span><span>"paris"</span><span>).</span><span>then</span><span>(</span><span>function</span><span>(</span><span>coords</span><span>)</span> <span>{</span> <span>console</span><span>.</span><span>log</span><span>(</span><span>coords</span><span>)</span> <span>});</span>
<span>// Object {lat: 48.8565056, lon: 2.3521334}</span>
</code></pre></div>
<p>Without promises, this would look like this:</p>
<div><pre><code data-lang="javascript"><span></span><span>var</span> <span>geocode</span> <span>=</span> <span>function</span><span>(</span><span>q</span><span>,</span> <span>cb</span><span>)</span> <span>{</span>
  <span>return</span> <span>$</span><span>.</span><span>ajax</span><span>(</span><span>"http://nominatim.openstreetmap.org/search/"</span><span>,</span> <span>{</span>
    <span>data</span><span>:</span> <span>{</span>
      <span>q</span><span>:</span> <span>q</span><span>,</span>
      <span>format</span><span>:</span> <span>"json"</span>
    <span>},</span>
    <span>success</span><span>:</span> <span>function</span><span>(</span><span>data</span><span>)</span> <span>{</span>
      <span>if</span> <span>(</span><span>!</span><span>data</span> <span>||</span> <span>!</span><span>data</span><span>[</span><span>0</span><span>])</span> <span>{</span>
        <span>throw</span> <span>new</span> <span>Error</span><span>(</span><span>"Geocoding '"</span> <span>+</span> <span>q</span> <span>+</span> <span>"' failed."</span><span>);</span>
      <span>}</span>
      <span>cb</span><span>({</span>
        <span>lat</span><span>:</span> <span>parseFloat</span><span>(</span><span>data</span><span>[</span><span>0</span><span>].</span><span>lat</span><span>,</span> <span>10</span><span>),</span>
        <span>lon</span><span>:</span> <span>parseFloat</span><span>(</span><span>data</span><span>[</span><span>0</span><span>].</span><span>lon</span><span>,</span> <span>10</span><span>)</span>
      <span>});</span>
    <span>}</span>
  <span>});</span>
<span>};</span>
</code></pre></div>
<h2>Serial Requests</h2>
<p>The second use case is to send a number of requests in series, using data
retrieved from one request as parameters to the next.</p>
<p>Let’s say you wanted to geocode a place name, then use the resulting coordinates
to perform a reverse geocode.</p>
<div><pre><code data-lang="javascript"><span></span><span>var</span> <span>biGeocode</span> <span>=</span> <span>function</span><span>(</span><span>q</span><span>)</span> <span>{</span>
  <span>var</span> <span>data</span> <span>=</span> <span>{};</span>
  <span>return</span> <span>geocode</span><span>(</span><span>q</span><span>).</span><span>then</span><span>(</span><span>function</span><span>(</span><span>coords</span><span>)</span> <span>{</span>
    <span>data</span><span>.</span><span>forward</span> <span>=</span> <span>coords</span><span>;</span>
    <span>return</span> <span>geocode</span><span>.</span><span>reverse</span><span>(</span><span>coords</span><span>);</span>
  <span>}).</span><span>then</span><span>(</span><span>function</span><span>(</span><span>description</span><span>)</span> <span>{</span>
    <span>data</span><span>.</span><span>reverse</span> <span>=</span> <span>description</span><span>;</span>
    <span>return</span> <span>data</span><span>;</span>
  <span>});</span>
<span>};</span>
</code></pre></div>
<p>This would then be usable like this:</p>
<div><pre><code data-lang="javascript"><span></span><span>biGeocode</span><span>(</span><span>"paris"</span><span>).</span><span>then</span><span>(</span><span>function</span><span>(</span><span>data</span><span>)</span> <span>{</span>
  <span>var</span> <span>fwd</span> <span>=</span> <span>data</span><span>.</span><span>forward</span><span>;</span>
  <span>var</span> <span>rev</span> <span>=</span> <span>data</span><span>.</span><span>reverse</span><span>;</span>
  <span>console</span><span>.</span><span>log</span><span>(</span>
    <span>"paris ("</span> <span>+</span> <span>fwd</span><span>.</span><span>lat</span> <span>+</span> <span>","</span> <span>+</span> <span>fwd</span><span>.</span><span>lon</span> <span>+</span> <span>") is in "</span> <span>+</span>
    <span>rev</span><span>.</span><span>address</span><span>.</span><span>country</span>
  <span>);</span>
<span>});</span>
<span>// paris (48.8565056,2.3521334) is in France</span>
</code></pre></div>
<p>The callback version for comparison:</p>
<div><pre><code data-lang="javascript"><span></span><span>var</span> <span>biGeocode</span> <span>=</span> <span>function</span><span>(</span><span>q</span><span>,</span> <span>cb</span><span>)</span> <span>{</span>
  <span>geocode</span><span>(</span><span>q</span><span>,</span> <span>function</span><span>(</span><span>coords</span><span>)</span> <span>{</span>
    <span>geocode</span><span>.</span><span>reverse</span><span>(</span><span>coords</span><span>,</span> <span>function</span><span>(</span><span>description</span><span>)</span> <span>{</span>
      <span>cb</span><span>({</span>
        <span>forward</span><span>:</span> <span>coords</span><span>,</span>
        <span>reverse</span><span>:</span> <span>description</span>
      <span>});</span>
    <span>));</span>
  <span>});</span>
<span>};</span>
</code></pre></div>
<h2>Parallel Requests</h2>
<p>If I’m trying to send a bunch of independent requests, it doesn’t make sense to
queue them up. I want to make as many requests in parallel as the browser will
let me. Let’s say for instance, I wanted to geocode a list of places. I can use
jQuery’s handy <a href="http://api.jquery.com/jquery.when/"><code>$.when</code></a>.</p>
<p>For a simple first example, let’s say I just wanted to do this once off, with a
known list of places.</p>
<div><pre><code data-lang="javascript"><span></span><span>$</span><span>.</span><span>when</span><span>(</span>
  <span>geocode</span><span>(</span><span>"Paris, France"</span><span>),</span>
  <span>geocode</span><span>(</span><span>"Toronto, Canada"</span><span>),</span>
  <span>geocode</span><span>(</span><span>"Waterloo, Canada"</span><span>)</span>
<span>).</span><span>then</span><span>(</span><span>function</span><span>(</span><span>parisCoords</span><span>,</span> <span>torontoCoords</span><span>,</span> <span>waterlooCoords</span><span>)</span> <span>{</span>
  <span>console</span><span>.</span><span>table</span><span>({</span>
    <span>"paris"</span><span>:</span> <span>parisCoords</span><span>,</span>
    <span>"toronto"</span><span>:</span> <span>torontoCoords</span><span>,</span>
    <span>"waterloo"</span><span>:</span> <span>waterlooCoords</span>
  <span>});</span>
<span>});</span>
</code></pre></div>
<p>And the callback version:</p>
<div><pre><code data-lang="javascript"><span></span><span>var</span> <span>pendingRequests</span> <span>=</span> <span>3</span><span>;</span>
<span>var</span> <span>results</span> <span>=</span> <span>{};</span>
<span>var</span> <span>parisCoords</span><span>,</span> <span>torontoCoords</span><span>,</span> <span>waterlooCoords</span><span>;</span>
<span>var</span> <span>done</span> <span>=</span> <span>function</span><span>()</span> <span>{</span>
  <span>if</span> <span>(</span><span>pendingRequests</span><span>--</span> <span>==</span> <span>0</span><span>)</span> <span>{</span>
    <span>console</span><span>.</span><span>table</span><span>({</span>
      <span>"paris"</span><span>:</span> <span>parisCoords</span><span>,</span>
      <span>"toronto"</span><span>:</span> <span>torontoCoords</span><span>,</span>
      <span>"waterloo"</span><span>:</span> <span>waterlooCoords</span>
    <span>});</span>
  <span>}</span>
<span>};</span>

<span>geocode</span><span>(</span><span>"Paris, France"</span><span>,</span> <span>function</span><span>(</span><span>coords</span><span>)</span> <span>{</span>
  <span>parisCoords</span> <span>=</span> <span>coords</span><span>;</span>
  <span>done</span><span>();</span>
<span>});</span>

<span>geocode</span><span>(</span><span>"Toronto, Canada"</span><span>,</span> <span>function</span><span>(</span><span>coords</span><span>)</span> <span>{</span>
  <span>torontoCoords</span> <span>=</span> <span>coords</span><span>;</span>
  <span>done</span><span>();</span>
<span>});</span>

<span>geocode</span><span>(</span><span>"Waterloo, Canada"</span><span>,</span> <span>function</span><span>(</span><span>coords</span><span>)</span> <span>{</span>
  <span>waterlooCoords</span> <span>=</span> <span>coords</span><span>;</span>
  <span>done</span><span>();</span>
<span>});</span>
</code></pre></div>
<p>Now let’s look at a more general case: geocoding an arbitrary list of places.</p>
<div><pre><code data-lang="javascript"><span></span><span>var</span> <span>batchGeocode</span> <span>=</span> <span>function</span><span>(</span><span>places</span><span>)</span> <span>{</span>
  <span>return</span> <span>$</span><span>.</span><span>when</span><span>.</span><span>apply</span><span>(</span><span>$</span><span>.</span><span>when</span><span>,</span> <span>places</span><span>.</span><span>map</span><span>(</span><span>geocode</span><span>)).</span><span>then</span><span>(</span><span>function</span><span>()</span> <span>{</span>
    <span>return</span> <span>Array</span><span>.</span><span>prototype</span><span>.</span><span>slice</span><span>.</span><span>apply</span><span>(</span><span>arguments</span><span>);</span>
  <span>});</span>
<span>};</span>
</code></pre></div>
<p>Which could then be used like this:</p>
<div><pre><code data-lang="javascript"><span></span><span>batchGeocode</span><span>([</span>
  <span>"Paris, France"</span><span>,</span>
  <span>"Toronto, Canada"</span><span>,</span>
  <span>"Waterloo, Canada"</span>
<span>]).</span><span>then</span><span>(</span><span>function</span><span>(</span><span>coords</span><span>)</span> <span>{</span>
  <span>console</span><span>.</span><span>table</span><span>({</span>
    <span>"paris"</span><span>:</span> <span>coords</span><span>[</span><span>0</span><span>],</span>
    <span>"toronto"</span><span>:</span> <span>coords</span><span>[</span><span>1</span><span>],</span>
    <span>"waterloo"</span><span>:</span> <span>coords</span><span>[</span><span>2</span><span>]</span>
  <span>});</span>
<span>});</span>
</code></pre></div>
<p>If the above looks arcane to you, you may wish to read about <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function/apply"><code>.apply</code></a>,
<a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/prototype"><code>.prototype</code></a>, and <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions_and_function_scope/arguments"><code>arguments</code></a>. The <code>Array.prototype.slice.apply</code> call
is to convert the <code>arguments</code> object into a proper <code>Array</code> instance. This also
assumes either your browser natively supports <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map"><code>.map</code></a> or you have
<a href="https://github.com/es-shims/es5-shim">es5-shim</a> on the page.</p>
<p>The equivalent code using callbacks:</p>
<div><pre><code data-lang="javascript"><span></span><span>var</span> <span>batchGeocode</span> <span>=</span> <span>function</span><span>(</span><span>places</span><span>,</span> <span>cb</span><span>)</span> <span>{</span>
  <span>var</span> <span>pendingRequests</span> <span>=</span> <span>places</span><span>.</span><span>length</span><span>;</span>
  <span>var</span> <span>result</span> <span>=</span> <span>places</span><span>.</span><span>map</span><span>(</span><span>function</span><span>()</span> <span>{</span> <span>return</span> <span>null</span><span>;</span> <span>});</span>
  <span>places</span><span>.</span><span>forEach</span><span>(</span><span>function</span><span>(</span><span>place</span><span>,</span> <span>i</span><span>)</span> <span>{</span>
    <span>geocode</span><span>(</span><span>place</span><span>,</span> <span>function</span><span>(</span><span>coords</span><span>)</span> <span>{</span>
      <span>result</span><span>[</span><span>i</span><span>]</span> <span>=</span> <span>coords</span><span>;</span>
      <span>if</span> <span>(</span><span>pendingRequests</span><span>--</span> <span>==</span> <span>0</span><span>)</span> <span>{</span>
        <span>cb</span><span>(</span><span>result</span><span>);</span>
      <span>}</span>
    <span>});</span>
  <span>});</span>
<span>};</span>
</code></pre></div>
<p>All of the interesting rendering is done using D3 to manipulate SVG. After
looking through a bunch of D3 examples, including the mesmerizing <a href="http://bl.ocks.org/mbostock/3711652">projection
transitions</a>, I settled on using Mike Bostock’s <a href="http://bl.ocks.org/mbostock/4180634">World Map bl.ock</a> as a
reference point. The key bit of magic for rendering the points on the map comes
from D3’s <a href="https://github.com/mbostock/d3/wiki/Geo-Projections">geo projections</a>.</p>
<h2>Projections</h2>
<p>Geo projections convert from a (longitude, latitude) pair to an (x, y)
coordinate given a kind of map projection and dimensions. I used the Mercator
projection because it’s so familiar, and because <a href="http://xkcd.com/977/">I’m not really that into
maps</a>.</p>
<p>Because of the beautiful interface of D3’s geo projections, I was able to
project the same data at several different granularities. The original version
showed only the world map, but after I loaded in all the data, it was impossible
to see any meaningful patterns in Europe or in the UK at that scale because of
the density. I ended up using 4 different projections, all Mercator, specified
by differing lat/lon bounding boxes.</p>
<p>You can see all the projections I used in <a href="https://github.com/jlfwong/travelmap/blob/master/app/projections.js"><code>app/projections.js</code></a>.</p>
<h2>Places and Paths</h2>
<p>Using these wonderful geo projections, I was able to lay down markers for all
the places we’d been and the paths we took to get there.</p>
<p>Each different color represents one of the people in my family. The colors were
selected using the D3’s categorical colors made available through
<a href="https://github.com/mbostock/d3/wiki/Ordinal-Scales#wiki-category10"><code>d3.scale.category10</code></a>.</p>
<p>Each different place is indicated by a circle or a pie slice, generated using
<a href="https://github.com/mbostock/d3/wiki/SVG-Shapes#wiki-arc"><code>d3.svg.arc</code></a>. If a person’s color is part of the pie, it means they’ve
been to that place. The more times anyone has visited the place, the larger the
pie, following a logarithmic scale. If I did this on a linear, or even square
root scale, the places my family members have called home (particularly my
hometown of Ottawa) would’ve ended up dominating the map far too much.</p>
<p>The arcs joining places represent the paths that people took. You can tell
which direction the path was in (i.e. differentiate between a path A → B
and B → A) by looking at the direction of curvature. An upward inflection
on the arc indicates travel East, downward inflection indicates travel West, and
similarly leftward for North and rightward for South. Because I wanted to make
commonly traveled back-and-forth routes apparent, I jitter the arc radius so
that each back-and-forth arc is slightly offset from previous ones. If you show
only me (Jamie) when viewing North America, you can clearly see my repeated
trips from Toronto to San Francisco and back for internships.</p>
<p>All of the pie slices for places and arcs for paths have a CSS class attached to
them matching the name of the person they correspond to. This makes hiding or
showing all of them a simple matter of selecting everything with that class and
setting their CSS <code>display</code> property.</p>
<h2>Animating</h2>
<p><img alt="travelmap" src="http://jamie-wong.com/images/14-01-03/pathanimation.gif"></p>
<p>To stagger the animation of all the places and paths, I used the <code>.transition</code>,
<code>.delay</code>, <code>.duration</code> triplet described very clearly in the <a href="http://chimera.labs.oreilly.com/books/1230000000345/ch09.html#_please_do_not_delay">“Please Do Not
Delay”</a> section of <a href="http://www.tkqlhce.com/click-7365820-11260198?url=http%3A%2F%2Fshop.oreilly.com%2Fproduct%2F0636920026938.do%3Fcmp%3Daf-strata-books-videos-product_cj_9781449339722_%2525zp&amp;cjsku=9781449339722">Interactive Data Visualization for the Web</a>.
The places animate over <code>fill-opacity</code>. The path arcs actually have a more
complex animation that you can see above that animates stretching out the arc
from the source to the destination. In the real version, however, the animation
happens so fast that you can’t really tell. Regardless of the number of paths
or arcs on each map, the total animation time is constant. This means the
animations on the world map are faster, because there are so many more places
and arcs as compared to the map of Europe.</p>
<p>The animations are triggered when the middle of the map hits the bottom of the
browser viewport as you scroll. The <a href="http://imakewebthings.com/jquery-waypoints/">jQuery Waypoints</a> plugin makes this
really easy. I use a slightly modified version of the same plugin to stick the
toggles in the top left of the screen only after you’ve scrolled past them.</p>
<p>When I’m working on these small <a href="http://jamie-wong.com/2013/05/05/something-out-of-nothing/">Something out of Nothing</a> projects of my
own, I like working with a static app, because it means deploying it is as
simple as uploading some assets to my VPS and letting nginx serve it. Having
done a few projects like this, I’ve temporarily settled on using <a href="http://brunch.io/">Brunch</a>
for developing and producing the final static assets.</p>
<p>This allows me to work with files that need preprocessing and allows me to use
the CommonJS-esque <code>require()</code> provided by Brunch (though I’ll likely be
switching to <a href="http://browserify.org/">Browserify</a> for the next project), and transpiled languages
like <a href="http://coffeescript.org/">CoffeeScript</a> and <a href="http://learnboost.github.io/stylus/">Stylus</a>. Because I’ve been working a lot more
in straight JavaScript than CoffeeScript recently, I decided to work without
CoffeeScript for this project, though it’s interesting to note that the chaining
syntax used extensively in jQuery and D3 just got sugary support in CoffeeScript
<em>two years after the proposal</em> in <a href="https://github.com/jashkenas/coffee-script/pull/3263">coffee-script#3263</a>.</p>
<p>To make this run completely serverless and to keep a good workflow between vim
and the browser, I decided to make all the geocoding requests from the
clientside. This is easy to do, but I wanted to avoid making thousands of AJAX
requests on page load, so I needed to cache the results somehow.</p>
<p><a href="http://lodash.com/">Lo-Dash</a> and <a href="http://underscorejs.org/">Underscore</a> both provide a <a href="http://lodash.com/docs#memoize"><code>_.memoize</code></a> function
which caches the results of expensive computations. This works great for
repeated operations within a single page session, but I wanted to cache to be
warm after a page reload, so I store the results in <a href="https://developer.mozilla.org/en-US/docs/Web/Guide/API/DOM/Storage#localStorage">localStorage</a>. I also
wanted to change the default behaviour of <code>_.memoize</code> only caching based on the
first argument.</p>
<p>After I coded away a solution for doing this with general functions, I realized
all of my use cases for it actually involved caching the results of promises,
not synchronous function calls. This requires special handling, because I don’t
want to cache the promise itself, I want to cache the value the promise resolved
to. The result was the <code>localStorageMemoize.promise</code> function you can see in
<a href="https://github.com/jlfwong/travelmap/blob/master/app/lib/localstorage_memoize.js#L28"><code>app/lib/localstorage_memoize.js</code></a>.</p>
<p>Given the promise-based <code>geocode</code> function discussed above, a cached version of
this looks like so:</p>
<div><pre><code data-lang="javascript"><span></span><span>var</span> <span>cachedGeocode</span> <span>=</span> <span>localStorageMemoize</span><span>.</span><span>promise</span><span>(</span><span>"geocoder"</span><span>,</span> <span>geocode</span><span>);</span>
<span>cachedGeocode</span><span>(</span><span>"paris"</span><span>).</span><span>then</span><span>(</span><span>function</span><span>(</span><span>coords</span><span>)</span> <span>{</span> <span>console</span><span>.</span><span>log</span><span>(</span><span>coords</span><span>);</span> <span>});</span>
</code></pre></div>
<p>On the first page load, the above will make the AJAX request, but on subsequent
page loads it’ll use the value cached in <code>localStorage</code>.</p>
<p>But what about deploying? Having all of these results cached in <code>localStorage</code>
is fine and dandy for developing locally, but you don’t want every visitor to
the page waiting for thousands of AJAX requests to finish on their first page
load. Having a speedy second load doesn’t matter if nobody sticks around long
enough for the first one to finish.</p>
<p>My solution was to provide the ability to pre-load the cache with a JSON blob.<br>
From my dev environment which already has everything cached, I download the
contents of the <code>localStorage</code> keys I’m interested in to disk using a clever
little script originally called <a href="http://bgrins.github.io/devtools-snippets/#console-save"><code>console.save</code></a>.</p>
<p>While this works and was an interesting exploration, it’s absolutely solving the
wrong problem. A much more sensible solution would be to just do all this
processing offline and produce a single JSON file containing all the data
post-geocoding and aggregation. This would’ve complicated my workflow a little
locally, but is definitely a more sensible solution in general.</p>

<p><img border="0" width="1" src="http://www.tqlkg.com/image-7365820-11260198" height="1"></p>
</article></div><br>
<br>
<br>
<br>
<br>
<br>
via Zero Wind :: Jamie Wong http://jamie-wong.com<br>
<br>
June 22, 2019 at 11:13PM
